README.md
# X4 Named Pipes API

Adds support for Windows named pipes to X4 Foundations.
Named pipes are OS level psuedo-files which support inter-process communication.
Pipes avoid the overhead of disk access from normal files (eg. debug logs), and are bidirectional.
X4 will act as a client, with one or more an external applications serving the pipes.

There are three components to this API:
 * A low level lua plugin and corresponding dll for pipe access.
 * An MD wrapper that interfaces between the lua and MD pipe users.
 * A default, optional Python external pipe host server.

### Requirements

* Lua Loader API extension
  - https://github.com/bvbohnen/x4-lua-loader-api
* Time API extension
  - https://github.com/bvbohnen/x4-time-api
* Optionally, Simple Menu API
  - https://github.com/bvbohnen/x4-simple-menu-api
* Optionally, Python 3.6+ with the pywin32 package.
  - Only needed if running the pipe server from python source code.
  - An executable is provided as an alternative.
  - The pywin32 package is part of the Anaconda distribution of python by default.

### Installation

* Place the named_pipes_api folder in extensions, as with normal mods.
* Place winpipe64.dll in ui\core\lualibs.
* Place the X4_Python_Pipe_Server (exe or py) anywhere convenient.
  - Run this pipe server alongside X4.

### Components

* MD Pipe API

  This api is found in Named_Pipes.xml, and allows MD code to read or write pipes.

* MD Server API

  This api is found in Pipe_Server_Host, and allows MD code to register a python module with the host server.
  The host (if running) will dynamically import the custom module.
  Such modules may be distributed with extensions.

* Lua Pipe API

  The lower level support for X4 to use named pipes is located in Named_Pipes.lua.
  Basic pipe usage can ignore this level in favor of the MD api.
  Advanced usage may want to use this api directly.

* Winpipe dll API

  The pipes themselves are accessed through a dll which wraps key Windows functions.
  This dll is limited to only what is needed: opening a pipe client, reading/writing/closing it, and error checking.
  If desired, the dll may be compiled freshly from source on github.

change_log.md

Change Log

* 0.1
  - Initial release.
  - Used the lua winapi module, compiled against x4 lua, to access windows
    named pipes.
  - Functional versions of lua and md apis in place, with light test code.
* 0.2
  - Rewrote winapi into winpipe, using only open_pipe and related file
    read/write/close functions.  Reduces dll size and improves security.
  - Switched client pipes into messaging mode, and verified pipelined reads
    are received correctly.
* 0.3
  - Refined behavior of api on connection errors and game reloads.
  - Rewrote lua api to support non-blocking access, with support for
    queueing multiple requests.  Winpipe updated accordingly.
  - Rewrote md api, compressing code, removing unused stubs.
  - Timeouts are a work in progress, and are disabled by default.
* 0.4
  - Timeouts working, plus some general debug and refinement.
* 0.5
  - Added workaround for lua garbage collection not closing pipe files.
  - General debug.
* 0.6
  - Slight debug to prevent excess garbage collection on pipes.
  - Reorganization of python test server setup, for scalability to
    running user provided subservers, and to support auto-restart
    when x4 reloads.
* 0.8
  - Added Pipe_Server_Lib.
  - Overhauled python server to work with the above.
* 0.9
  - Added the Server_Reader library function which handles connecting
    to and continually reading a pipe, and updated Pipe_Server_Lib.
* 0.9.1
  - Minor tweaks.
* 0.10
  - Added status report menu using simple menu api.
* 0.11
  - Switch to the Time API setting time between reconnect pings while the game is paused.
* 0.12
  - Exposed lua api to other lua modules.
* 0.13
  - Fixed x4 crash when using /reloadui that showed up in 3.0 beta.
  - Reduced reliance on onUpdate events for pipe reading, due to dropouts.
* 0.14
  - dll file moved inside the extension, and no longer needs to be in ui/core/lualibs.
* 1.0
  - General release.
* 1.0.1
  - Removed unnecessary dependencies from the winpipe dll.
* 1.1
  - Added $cancel_on_timeout arg to the Access_Handler.
  - Changed default timeout to essentially be disabled.

X4_Files/Change_Log.py
'''
Change Log

* 0.1
  - Initial release.
  - Used the lua winapi module, compiled against x4 lua, to access windows
    named pipes.
  - Functional versions of lua and md apis in place, with light test code.
* 0.2
  - Rewrote winapi into winpipe, using only open_pipe and related file
    read/write/close functions.  Reduces dll size and improves security.
  - Switched client pipes into messaging mode, and verified pipelined reads
    are received correctly.
  - Added Make_Release for zip file generation with the X4 files.
* 0.3
  - Refined behavior of api on connection errors and game reloads.
  - Rewrote lua api to support non-blocking access, with support for
    queueing multiple requests.  Winpipe updated accordingly.
  - Rewrote md api, compressing code, removing unused stubs.
  - Timeouts are a work in progress, and are disabled by default.
* 0.4
  - Timeouts working, plus some general debug and refinement.
* 0.5
  - Added workaround for lua garbage collection not closing pipe files.
  - General debug.
* 0.6
  - Slight debug to prevent excess garbage collection on pipes.
  - Reorganization of python test server setup, for scalability to
    running user provided subservers, and to support auto-restart
    when x4 reloads.
* 0.7
  - Added key_capture_api and related python server code, for capturing
    key/combo presses and linking them to cues.
* 0.71
  - Debug on key capture api.
* 0.8
  - Added lua_loader_api.
  - Added LIB_Pipe_Server.
  - Overhauled python server to work with the above.
* 0.9
  - Added the Server_Reader library function which handles connecting
    to and continually reading a pipe, and updated LIB_Pipe_Server and
    LIB_Key_Capture to use it.
  - Key capture api now requires keys be re-registered when it reloads.
'''

def Get_Version():
    '''
    Returns the highest version number in the change log,
    as a string, eg. '3.4.1'.
    '''
    # Traverse the docstring, looking for ' *' lines, and keep recording
    #  strings as they are seen.
    version = ''
    for line in __doc__.splitlines():
        if not line.startswith('*'):
            continue
        version = line.split('*')[1].strip()
    return version
